---
title: "global_proteomics_NMF_clustering"
output: html_document
date: "2024-10-06"
---

```{r}
suppressMessages(library(preprocessCore))
suppressMessages(library(CancerSubtypes))
suppressMessages(library(missForest))
suppressMessages(library(ComplexHeatmap))
suppressMessages(library(ggplot2))
suppressMessages(library(circlize))
suppressMessages(library(colorRamp2))
```

### Read input file (proteomics)
```{r}
# Read global proteomics pre-processed input files
glob_p.df <- read.delim("/Users/onco/Downloads/LAIDD_GASTRIC_R/Proteomics/Proteomics/protein_expression_change.txt", header=TRUE, sep = "\t", quote=NULL)
#glob_p.df <- read.delim("C:/Users/YounsooKang/Downloads/LAIDD_GASTRIC_R/Proteomics/Proteomics/protein_expression_change.txt", header=TRUE, sep = "\t", quote=NULL)
```

### Filter NA
```{r}
# Filter only the rows where more than 50% of the values are not NA.
glob_p.df.f <- glob_p.df[rowMeans(!is.na(glob_p.df[, 7:ncol(glob_p.df)])) >= 0.5, ]

dim(glob_p.df)
dim(glob_p.df.f)
```

### Quantile Normalization
```{r}
# Separate gene information
prot_info <- glob_p.df.f[, 1:6]  
glob_p.data.mat <- as.matrix(glob_p.df.f[, 7:ncol(glob_p.df.f)])  

# Convert NaN into NA
glob_p.data.mat[is.nan(glob_p.data.mat)]<- NA

# Perform quantile normalization
library(preprocessCore)
norm.glob_p.data <- normalize.quantiles(glob_p.data.mat)

# Convert normalized data back to a dataframe and combine with gene information
norm.glob_p.df <- as.data.frame(norm.glob_p.data)
colnames(norm.glob_p.df) <- colnames(glob_p.df.f)[7:ncol(glob_p.df.f)] 

# Combine gene information with normalized data
final.glob_p.df <- cbind(prot_info, norm.glob_p.df)

# Compare data distribution
data.checkDistribution(glob_p.df.f[7:ncol(glob_p.df.f)])
data.checkDistribution(final.glob_p.df[7:ncol(final.glob_p.df)])
```

## Select one of the missing value imputation method

### Impute missing values using missForest (method1)
```{r}
library(missForest)

# impute missing values using random forest based algorithm
imputed_data <- missForest(final.glob_p.df[,7:ncol(final.glob_p.df)])

# imputed dataframe stored in imputed_data$ximp
final.glob_p.df.filt <- imputed_data$ximp
final.glob_p.df.filt <- cbind(final.glob_p.df[,1:6], final.glob_p.df.filt)

NRMSE <- imputed_data$OOBerror

cat("NRMSE (Normalized Root Mean Squared Error):", NRMSE, "\n")

# result
head(final.glob_p.df.filt)
```

### Impute missing values using mice (method2)
```{r}
library(mice)

imputed_data <- mice(final.glob_p.df[,7:ncol(final.glob_p.df)], m = 5, maxit = 50, meth = 'pmm', seed = 123)

final.glob_p.df.filt <- cbind(final.glob_p.df[,1:6], complete(imputed_data))

# save intermediate result
#write.table(final.glob_p.df.filt, "/Users/onco/Downloads/LAIDD_GASTRIC_R/Proteomics/Proteomics/imputed_p_change.txt", row.names=F, col.names=T, sep="\t", quote=FALSE)
```

```{r}
# load intermediate result
final.glob_p.df.filt <- read.delim("/Users/onco/Downloads/LAIDD_GASTRIC_R/Proteomics/Proteomics/imputed_p_change.txt", header=T, sep="\t", quote=NULL)
#final.glob_p.df.filt <- read.delim("C:/Users/YounsooKang/Downloads/LAIDD_GASTRIC_R/Proteomics/Proteomics/imputed_p_change.txt", header=T, sep="\t", quote=NULL)
```


### Top variant peptides
```{r}
# Calculate median absolute deviation
mad_values <- apply(final.glob_p.df.filt[7:ncol(final.glob_p.df.filt)], 1, mad)

# Subset variant peptides

# count
num_genes <- nrow(final.glob_p.df.filt)

# top 10% variant peptides
top_10_percent_threshold <- quantile(mad_values, 0.90)
top_10_df <- final.glob_p.df.filt[mad_values >= top_10_percent_threshold, ]

# top 20% variant peptides
top_20_percent_threshold <- quantile(mad_values, 0.80)
top_20_df <- final.glob_p.df.filt[mad_values >= top_20_percent_threshold, ]

# top 30% variant peptides
top_30_percent_threshold <- quantile(mad_values, 0.70)
top_30_df <- final.glob_p.df.filt[mad_values >= top_30_percent_threshold, ]

# result
dim(top_10_df)  
dim(top_20_df)  
dim(top_30_df)  

```

### Decoupling 
```{r}
# replace negative to zero
replace_negative <- function(x) {
  x[x < 0] <- 0
  return(as.matrix(x))
}

# only positives
positive_10 <- replace_negative(top_10_df[7:ncol(top_10_df)])
positive_20 <- replace_negative(top_20_df[7:ncol(top_20_df)])
positive_30 <- replace_negative(top_30_df[7:ncol(top_30_df)])


# only negatives
negative_10 <- replace_negative(-top_10_df[7:ncol(top_10_df)])
negative_20 <- replace_negative(-top_20_df[7:ncol(top_20_df)])
negative_30 <- replace_negative(-top_30_df[7:ncol(top_30_df)])

# merge
merge_10 <- list(positive_10, negative_10)
merge_20 <- list(positive_20, negative_20)
merge_30 <- list(positive_30, negative_30)
```

## consensus NMF clustering from CancerSubtypes package
### Cophenetic correlation plot

```{r}
# Cluster range
cluster_range <- 2:6
#cluster_range=4

# Save results
results <- list()

cophenetic_corr_values <- c()

# cophenetic correlation 값을 저장할 리스트
cophenetic_corr_values_list <- list()

# 각 merge 데이터에 대해 반복
for (percentage in c("10%", "20%", "30%")) {
  
  # merge 데이터 설정
  merge_data <- switch(percentage,
                       "10%" = merge_10,
                       "20%" = merge_20,
                       "30%" = merge_30)
  
  # cophenetic correlation 값을 저장할 벡터 초기화
  cophenetic_corr_values <- numeric(length(cluster_range))
  
  # 각 클러스터에 대해 CNMF 실행
  for (k in cluster_range) {
    cat("Running CNMF with", k, "clusters on", percentage, "data...\n")
    
    # CNMF 실행 결과 저장
    results[[paste0("Cluster_", k, "_", percentage)]] <- ExecuteCNMF(merge_data, clusterNum = k, nrun = 200)
    distance_matrix <- results[[paste0("Cluster_", k, "_", percentage)]]$distanceMatrix
    
    # Cophenetic correlation 계산
    d1 <- dist(distance_matrix)
    hclust_obj <- hclust(d1, method = "average")
    d2 <- cophenetic(hclust_obj)
    cophenetic_corr <- cor(d1, d2)
    
    # cophenetic correlation 값을 벡터에 저장
    cophenetic_corr_values[k - min(cluster_range) + 1] <- cophenetic_corr
    
    cat("Cophenetic correlation for", k, "clusters:", cophenetic_corr, "\n")
  }
  
  # 결과를 리스트에 저장
  cophenetic_corr_values_list[[percentage]] <- cophenetic_corr_values
}

# 데이터 프레임 생성
data <- data.frame(
  Clusters = rep(cluster_range, 3),
  Cophenetic = unlist(cophenetic_corr_values_list),
  Percentage = rep(c("10%", "20%", "30%"), each = length(cluster_range))
)

# 결과 출력
print(data)
```

# silhouette plots
```{r}
# Generate silhouette plots for each percentage data
for (percentage in c("10%", "20%", "30%")) {
    for (i in cluster_range) {
        # Retrieve group and distance matrix for each cluster and percentage
        result_key <- paste0("Cluster_", i, "_", percentage)
        
        # Check if the result exists
        if (!is.null(results[[result_key]])) {
            cluster_group <- results[[result_key]]$group
            distance_matrix <- results[[result_key]]$distanceMatrix
            
            # Calculate silhouette
            if (!is.null(cluster_group) && !is.null(distance_matrix)) {
                sil_result <- silhouette_SimilarityMatrix(cluster_group, distance_matrix)
                
                # Generate plot
                plot(sil_result, main = paste("Silhouette Plot for", i, "Clusters on", percentage, "Data"))
            } else {
                cat("Skipping cluster", i, "with", percentage, "data: NULL values detected in group or distanceMatrix\n")
            }
        } else {
            cat("No result found for", result_key, "\n")
        }
    }
}
```


### Final cNMF results
```{r}
# select the best results for MAD=20, k=4
cluster_group20_4 <- results$`Cluster_4_20%`$group
distance_matrix20_4 <- results$`Cluster_4_20%`$distanceMatrix
sil20_result_4 <- silhouette_SimilarityMatrix(cluster_group20_4, distance_matrix20_4)
Consensus20_4 <- results$`Cluster_4_20%`$originalResult@consensus
```


### plots
```{r}
# Heatmap
group_info <- paste0("Prot", cluster_group20_4)
group_colors <- c("Prot1" = "orange", "Prot2" = "green", "Prot3" = "purple", "Prot4" = "red")

# Bottom annotation 생성
bottom_annotation <- HeatmapAnnotation(
  group = group_info, 
  col = list(group = group_colors),
  annotation_legend_param = list(title = "group"),
  annotation_height = unit(4, "mm"),
  show_legend = FALSE
)

NMF_heatmap <- Heatmap(Consensus20_4, 
                      name = "Consensus value", 
                      show_column_names = FALSE, 
                      show_row_names = FALSE,
                      bottom_annotation = bottom_annotation)
NMF_heatmap


# Cophenetic plot
cophenetic_plot <- ggplot(data, aes(x = Clusters, y = Cophenetic, color = Percentage, group = Percentage)) +
                          geom_line(size = 1) +                       
                          geom_point(size = 2) +                      
                          labs(title = "Cophenetic Correlation vs Number of Clusters", 
                               x = "Number of clusters", 
                               y = "Cophenetic correlation") + 
                          theme_minimal() +                           
                          theme(plot.title = element_text(hjust = 0.5)) + 
                          scale_color_manual(values = c("green", "blue", "red"))  

cophenetic_plot



# Silhouette width plot
sil_plot <- plot(sil20_result_4, col="darkblue")
# Extract Silhouette values
sil_values <- sil20_result_4[, "sil_width"]  
sorted_sil_values <- sort(sil_values)
silhouette_plot <- barplot(sorted_sil_values, horiz = TRUE, 
                              col = "darkblue",
                              border = TRUE,
                              xlab = "Silhouette width", 
                              ylab = "Sorted samples", 
                              xlim = c(0, max(sorted_sil_values)))
silhouette_plot
```


### Save plots
```{r}
#pdf("C:/Users/YounsooKang/Downloads/LAIDD_GASTRIC_R/Proteomics/Proteomics/globp_heatmap_output.pdf", width = 12, height = 10)
pdf("/Users/onco/Downloads/LAIDD_GASTRIC_R/Proteomics/Proteomics/globp_heatmap_output.pdf", width = 12, height = 10)
NMF_heatmap
dev.off()

#pdf("C:/Users/YounsooKang/Downloads/LAIDD_GASTRIC_R/Proteomics/Proteomics/globp_silhouette_barplot.pdf", width = 12, height = 10)
pdf("/Users/onco/Downloads/LAIDD_GASTRIC_R/Proteomics/Proteomics/globp_silhouette_barplot.pdf", width = 12, height = 10)
silhouette_plot
dev.off()

#pdf("C:/Users/YounsooKang/Downloads/LAIDD_GASTRIC_R/Proteomics/Proteomics/globp_cophenetic_coef.pdf", width = 12, height = 10)
pdf("/Users/onco/Downloads/LAIDD_GASTRIC_R/Proteomics/Proteomics/globp_cophenetic_coef.pdf", width = 12, height = 10)
cophenetic_plot
dev.off()
```

### Subtype Signature Molecules
```{r}
# Set sample name in the first column
rownames(sil20_result_4) <- colnames(top_20_df)[7:ncol(top_20_df)]

# Select samples with silhouette width > 0
positive_sil_samples <- rownames(sil20_result_4)[sil20_result_4[, 'sil_width'] > 0]
negative_sil_samples <- rownames(sil20_result_4)[sil20_result_4[, 'sil_width'] < 0]

# positive silhouette matrix
sil20_result_4_pos <- sil20_result_4[positive_sil_samples,]
# sil value 0.03 제외 (too low)
sil20_result_4_pos <- sil20_result_4_pos[rownames(sil20_result_4_pos) != "N219T220", ]

dim(sil20_result_4_pos)

# Filter the original data frame to keep only the samples with positive silhouette scores
positive_sample_df <- top_20_df[, c("IND", "Full.ID", "Prot_ID", "X..of.value", "Symbol", "EntrezID", positive_sil_samples)]

# data
top_20_df_pos <- top_20_df[, c("IND", "Full.ID", "Prot_ID", "X..of.value", "Symbol", "EntrezID", positive_sil_samples)]
```

```{r}
rownames(top_20_df_pos) <- top_20_df_pos$Prot_ID
rownames(top_20_df) <- top_20_df$Prot_ID
```


### DEP
```{r}
# Perform t-tests to identify DEGs for each cluster with the additional fold change condition
find_deg_with_fc <- function(cluster, data, p_thresh = 0.05, fc_thresh = 2) {
  
  target_cluster <- rownames(sil20_result_4_pos)[sil20_result_4_pos[, 'cluster'] == cluster]
  other_clusters <- rownames(sil20_result_4_pos)[sil20_result_4_pos[, 'cluster'] != cluster]
  
  target_df <- data[, target_cluster]  # target cluster
  other_df <- data[, other_clusters]  # other clusters
  
  # Perform t-test and calculate fold change for each gene
  deg_genes <- apply(data, 1, function(gene_expr) {
    # Perform t-test
    t_test <- t.test(gene_expr[target_cluster], gene_expr[other_clusters])
    
    # Calculate mean values
    mean_target_cluster <- mean(gene_expr[target_cluster])
    mean_other_clusters <- mean(gene_expr[other_clusters])
    
    # Calculate fold change
    fold_change <- mean_target_cluster  / mean_other_clusters
    
    # Calculate median values
    median_target_cluster <- median(gene_expr[target_cluster])
    median_other_clusters <- median(gene_expr[other_clusters])
    
    # Only return genes where p-value is less than threshold and fold change is greater than or equal to threshold
    if (t_test$p.value < p_thresh && 
        #fold_change >= fc_thresh && 
        median_target_cluster > 0 && 
        median_other_clusters < 0 &&  
        median_target_cluster > median_other_clusters) {  
      return(TRUE)
    } else {
      return(FALSE)
    }
  })
  
  return(rownames(data)[deg_genes])
}
```


```{r}
# Find DEGs with the additional fold change filter for each cluster
deg_cluster1 <- find_deg_with_fc(1, top_20_df_pos[7:ncol(top_20_df_pos)])
deg_cluster2 <- find_deg_with_fc(2, top_20_df_pos[7:ncol(top_20_df_pos)])
deg_cluster3 <- find_deg_with_fc(3, top_20_df_pos[7:ncol(top_20_df_pos)])
deg_cluster4 <- find_deg_with_fc(4, top_20_df_pos[7:ncol(top_20_df_pos)])

# DEPs for each cluster
length(deg_cluster1) # 94
length(deg_cluster2) # 283
length(deg_cluster3) # 186
length(deg_cluster4) # 237

# Samples for each cluster
cluster1_samples <- rownames(sil20_result_4_pos)[sil20_result_4_pos[, 'cluster'] == 1]
cluster2_samples <- rownames(sil20_result_4_pos)[sil20_result_4_pos[, 'cluster'] == 2]
cluster3_samples <- rownames(sil20_result_4_pos)[sil20_result_4_pos[, 'cluster'] == 3]
cluster4_samples <- rownames(sil20_result_4_pos)[sil20_result_4_pos[, 'cluster'] == 4]

length(cluster1_samples) # 12
length(cluster2_samples) # 15
length(cluster3_samples) # 32
length(cluster4_samples) # 20

sort(c(length(cluster1_samples), length(cluster2_samples), length(cluster3_samples), length(cluster4_samples)), decreasing=TRUE)


deg_df <- top_20_df_pos[c(deg_cluster1, deg_cluster2, deg_cluster3, deg_cluster4), c(cluster1_samples, cluster2_samples, cluster3_samples, cluster4_samples)]

# gene cluster
gene_clusters <- factor(c(rep("Prot1", length(deg_cluster1)), rep("Prot2", length(deg_cluster2)), rep("Prot3", length(deg_cluster3)), rep("Prot4", length(deg_cluster4))))
names(gene_clusters) <- rownames(as.matrix(deg_df)) 

# sample cluster
sample_clusters <- factor(c(rep("Prot1", length(cluster1_samples)), rep("Prot2", length(cluster2_samples)), rep("Prot3", length(cluster3_samples)), rep("Prot4", length(cluster4_samples))))
names(sample_clusters) <- colnames(as.matrix(deg_df)) 

```


### signature heatmap
```{r}
# color
col_fun = colorRamp2(c(-2, 0, 2), c("green", "black", "red"))

# bottom annotation
column_ha <- HeatmapAnnotation(
  clusters = sample_clusters, 
  col = list(clusters = c("Prot1" = "#87CEEB", "Prot2" = "#DAA520", "Prot3" = "#FF00FF", "Prot4" = "#00FF00")),
  annotation_legend_param = list(clusters = list(title = "")),
  show_legend = FALSE,
  show_annotation_name = FALSE
)

# right annotation
row_ha <- rowAnnotation(
  clusters = gene_clusters, 
  col = list(clusters = c("Prot1" = "#87CEEB", "Prot2" = "#DAA520", "Prot3" = "#FF00FF", "Prot4" = "#00FF00")),
  annotation_legend_param = list(clusters = list(title = "")),
  show_legend = TRUE,
  show_annotation_name = FALSE
)


# signature heatmap
subtype_signature_heatmap <- Heatmap(
  as.matrix(deg_df), 
  name = "Abundance", 
  col = col_fun,
  bottom_annotation = column_ha,  
  right_annotation = row_ha,  
  cluster_rows = FALSE,  
  cluster_columns = FALSE,  
  show_row_names = FALSE,  
  show_column_names = FALSE
)

subtype_signature_heatmap
```

### Consensus heatmap
```{r}
# positive consensus matrix
Consensus20_4_pos <- Consensus20_4[positive_sil_samples, positive_sil_samples]

# exclude
#samples_to_exclude <- "N219T220"

# exclude
#Consensus20_4_pos <- Consensus20_4[!(rownames(Consensus20_4) %in% samples_to_exclude), !(colnames(Consensus20_4) %in% samples_to_exclude)]

# group info & colors
group_info_pos <- paste0("Prot", sil20_result_4_pos[,"cluster"])
group_colors_pos <- c("Prot1" = "#87CEEB", "Prot2" = "#DAA520", "Prot3" = "#FF00FF", "Prot4" = "#00FF00")

# Bottom annotation
bottom_annotation <- HeatmapAnnotation(
  group = group_info_pos, 
  col = list(group = group_colors_pos),
  annotation_legend_param = list(title = ""),
  annotation_height = unit(4, "mm"),
  show_legend = TRUE
)

# Consensus heatmap
Consensus_heatmap <- Heatmap(Consensus20_4_pos, 
                      name = "Consensus value", 
                      show_column_names = FALSE, 
                      show_row_names = FALSE,
                      bottom_annotation = bottom_annotation)

Consensus_heatmap
```

```{r}
pdf("/Users/Onco/Downloads/LAIDD_GASTRIC_R/Proteomics/Proteomics/signature_heatmap_globp.pdf", width = 12, height = 10)
subtype_signature_heatmap
dev.off()

pdf("/Users/Onco/Downloads/LAIDD_GASTRIC_R/Proteomics/Proteomics/consensus_heatmap_globp.pdf", width = 12, height = 10)
Consensus_heatmap
dev.off()
```

### Generate input matrix for integrative consensus clustering 
```{r}
sil20_result_4

# sil_matrix transponse
t_sil_matrix <- t(sil20_result_4)

# rownames
colnames(t_sil_matrix) <- rownames(sil20_result_4)

# Construct matrix
t_sil_matrix <- t_sil_matrix[c(1,2), ]
t_sil_matrix <- rbind(t_sil_matrix, NA)
t_sil_matrix <- rbind(t_sil_matrix, NA)

t_sil_matrix[2, ] <- t_sil_matrix[1, ]
t_sil_matrix[3, ] <- t_sil_matrix[1, ]
t_sil_matrix[4, ] <- t_sil_matrix[1, ]

# each row corresponds to identified cluster type
t_sil_matrix[1, ] <- ifelse(t_sil_matrix[1, ] == 1, 1, 0)
t_sil_matrix[2, ] <- ifelse(t_sil_matrix[2, ] == 2, 1, 0)
t_sil_matrix[3, ] <- ifelse(t_sil_matrix[3, ] == 3, 1, 0)
t_sil_matrix[4, ] <- ifelse(t_sil_matrix[4, ] == 4, 1, 0)

# numeric
t_sil_matrix <- apply(t_sil_matrix, 2, as.numeric)

# subtype names
icluster_index_glob_p <- t_sil_matrix
rownames(icluster_index_glob_p)[1:4] <- c("Prot1", "Prot2", "Prot3", "Prot4")

# save as csv
write.csv(icluster_index_glob_p, "icluster_index_p.csv", row.names = TRUE)
```

