{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1ef5bf72-e919-4ce1-b4c2-1951e33e94df",
   "metadata": {},
   "source": [
    "# 01_Patients_Analysis\n",
    "### --------------------\n",
    "### --------------------\n",
    "### --------------------"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48a43a99-d8dd-4ac7-bfe6-013812b45e56",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "pd.set_option('future.no_silent_downcasting', True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60cc51e6-7856-4257-bfd3-43fdf19485d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Loading patient metadata\n",
    "patient_meta = pd.read_csv(\"metadata.txt\",sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee8da1ec-1347-421e-ac9d-a5a83b126510",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Data processing\n",
    "patient_meta[\"Normal\"]=\"N\"+patient_meta[\"Normal\"].str[:-1]\n",
    "patient_meta[\"Tumor\"]=\"T\"+patient_meta[\"Tumor\"].str[:-1]\n",
    "patient_meta[\"sampleID\"]=patient_meta[\"Normal\"]+patient_meta[\"Tumor\"]\n",
    "patient_meta_selected = patient_meta[[\"sampleID\",\"EBV\",\"MSI\",\"Gender\",\"Histology (Lauren)\"]]\n",
    "patient_meta_selected = patient_meta_selected.rename(columns={\"Histology (Lauren)\":\"Histology\"})\n",
    "patient_meta_selected['Histology'] = patient_meta_selected['Histology'].fillna(\"Others\")\n",
    "patient_meta_selected['EBV'] = patient_meta_selected['EBV'].replace(\"Negative\", \"EBV-\")\n",
    "patient_meta_selected['EBV'] = patient_meta_selected['EBV'].replace(\"EBV(PIK3CAmut)\", \"EBV+\")\n",
    "patient_meta_selected['EBV'] = patient_meta_selected['EBV'].replace(\"EBV\", \"EBV+\")\n",
    "patient_meta_selected['EBV'] = patient_meta_selected['EBV'].replace(\"EBV(PIK3CAwt)\", \"EBV+\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e11dcf77-036e-48aa-bb82-17c371c22a7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Loading variant data\n",
    "df = pd.read_excel(\"WES_somaticVariants.xlsx\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50ca16e5-5735-4b9f-91a1-e6cbf80e3c94",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Data processing\n",
    "new_drop= new.drop_duplicates()\n",
    "new_group = new.groupby(['patientnum']).agg(lambda x: x.dropna().unique())\n",
    "no_mutation = list(set(patient_meta.sampleID.unique()) - set(new_group.index.unique()))\n",
    "no_mut = pd.DataFrame(no_mutation,columns=[\"patientnum\"])\n",
    "df_fillna = new_group.apply(lambda col: col.map(lambda x: np.nan if len(x) == 0 else x))\n",
    "df_fillna.reset_index(inplace=True)\n",
    "df_fillna[\"num\"] = df_fillna['patientnum'].str.extract(r'N(\\d+)T').astype(int)\n",
    "df_fillna = df_fillna.sort_values(by=\"num\").drop(\"num\",axis=1).set_index(\"patientnum\")\n",
    "df_fillna = df_fillna.apply(lambda col: col.map(lambda x: str(x)[1:-1] if pd.notnull(x) else x))\n",
    "df_fillna = df_fillna.apply(lambda col: col.map(lambda x: str(x).strip(\"'\") if pd.notnull(x) else x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6321a612-17bf-413f-9980-241defc587bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# mutation type\n",
    "data_map = {'frameshift deletion':\"Frame shift\",\n",
    " 'frameshift insertion':\"Frame shift\",\n",
    " 'nonframeshift deletion':\"In frame Indel\",\n",
    " 'nonsynonymous SNV':\"Missense\",\n",
    " 'splicing':\"Splice site\",\n",
    " 'stopgain SNV':\"Nonsense\"}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bcbe4a3a-3fa7-4398-aaf1-b2d677ae4318",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Data sorting\n",
    "df_fillna.reset_index(inplace=True)\n",
    "final_output = df_fillna.apply(lambda x: x.apply(lambda val: data_map.get(val, val)))\n",
    "final_output_sort = final_output.sort_values(by=\"RHOA\").sort_values(by=\"BANP\").sort_values(by=\"ARID1A\").sort_values(by=\"MUC5B\").sort_values(by=\"TP53\").sort_values(by=\"CDH1\")\n",
    "final_output_sort = final_output_sort[[\"patientnum\",\"CDH1\",\"TP53\",\"MUC5B\",\"ARID1A\",\"BANP\",\"RHOA\"]]\n",
    "final_output_sort = pd.concat([final_output_sort, no_mut],axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d456393c-2ca3-408b-90e8-1cd5db603fad",
   "metadata": {},
   "outputs": [],
   "source": [
    "# save\n",
    "final_output_sort.fillna(0).to_csv(\"mutationtype.csv\",index=False)\n",
    "patient_meta_selected.set_index('sampleID').loc[final_output_sort['patientnum']].reset_index().to_csv(\"patientmeta.csv\",index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0b2c99c-06c5-4a26-ab1b-6a995d50daea",
   "metadata": {},
   "source": [
    "# 02_Phosphopeptides_Analysis\n",
    "### --------------------\n",
    "### --------------------\n",
    "### --------------------"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f0d80cb-ed40-4fb6-a998-ad5271774a06",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from scipy import stats\n",
    "import qnorm\n",
    "from scipy.stats import wilcoxon\n",
    "from scipy.stats import mannwhitneyu"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e783070f-ff14-4ae8-90cd-f192afd72325",
   "metadata": {},
   "outputs": [],
   "source": [
    "#input data\n",
    "df = pd.read_csv(\"phosphopeptide_change.txt\",sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8184404-757b-4013-b9b9-68f331c686ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Consensuspathdb input data\n",
    "\n",
    "peptide_symbol = df[[\"Peptide\",\"Symbol\"]]\n",
    "\n",
    "arid1a_peptide = pd.DataFrame(arid1a_final_scaler.reset_index(),columns=[\"Peptide\"])\n",
    "pd.merge(peptide_symbol, arid1a_peptide, on=\"Peptide\",how=\"inner\").Symbol.to_csv(\"arid1a_symbol_phospho.csv\",index=False)\n",
    "\n",
    "CDH1_peptide = pd.DataFrame(CDH1_final_scaler.reset_index(),columns=[\"Peptide\"])\n",
    "pd.merge(peptide_symbol, CDH1_peptide, on=\"Peptide\",how=\"inner\").Symbol.to_csv(\"CDH1_symbol_phospho.csv\",index=False)\n",
    "\n",
    "RHOA_peptide = pd.DataFrame(RHOA_final_scaler.reset_index(),columns=[\"Peptide\"])\n",
    "pd.merge(peptide_symbol, RHOA_peptide, on=\"Peptide\",how=\"inner\").Symbol.to_csv(\"RHOA_symbol_phospho.csv\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f534fe4f-3449-4ca5-a918-e77d4e5f6fae",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Data processing for phosphopeptide signature\n",
    "\n",
    "#컬럼에 patient id만 남기도록 정리\n",
    "df = df.set_index(\"Peptide\")\n",
    "df.drop(\"Symbol\",axis=1,inplace=True)\n",
    "\n",
    "# NaN이 아닌 값이 50% 이상인 행만 선정\n",
    "selected_rows = df[df.count(axis=1) / len(df.columns) >= 0.5]\n",
    "\n",
    "# 새로운 데이터프레임 생성\n",
    "phospho_fch = df.loc[selected_rows.index]\n",
    "\n",
    "#quantile normalization\n",
    "phospho_fch_t = phospho_fch.transpose()\n",
    "df =qnorm.quantile_normalize(phospho_fch_t).transpose()\n",
    "\n",
    "mutation = pd.read_excel(\"SuppleTable_SomaticMutation_80_nonsyn.xls\")\n",
    "\n",
    "#각 유전자에 돌연변이가 존재하는 patient id 가져오기\n",
    "ARID1A_patient_id = mutation[mutation.Gene == \"ARID1A\"].Patient.to_list()\n",
    "RHOA_patient_id = mutation[mutation.Gene == \"RHOA\"].Patient.to_list()\n",
    "CDH1_patient_id = mutation[mutation.Gene == \"CDH1\"].Patient.to_list()\n",
    "\n",
    "#df와 mutation의 patient id 형식 맞춰주기\n",
    "arid1a = []\n",
    "for item in ARID1A_patient_id:\n",
    "    number_part, letter_part = item[:-1], item[-1]\n",
    "    arid1a.append(f'N{int(number_part) - 1}{letter_part}{number_part}')\n",
    "\n",
    "#df와 mutation의 patient id 형식 맞춰주기\n",
    "RHOA = []\n",
    "for item in RHOA_patient_id:\n",
    "    number_part, letter_part = item[:-1], item[-1]\n",
    "    RHOA.append(f'N{int(number_part) - 1}{letter_part}{number_part}')\n",
    "\n",
    "#df와 mutation의 patient id 형식 맞춰주기\n",
    "CDH1 = []\n",
    "for item in CDH1_patient_id:\n",
    "    number_part, letter_part = item[:-1], item[-1]\n",
    "    CDH1.append(f'N{int(number_part) - 1}{letter_part}{number_part}')\n",
    "\n",
    "#mutation과 wildtype 환자군 분류\n",
    "arid1a_mutated = df.columns[df.columns.isin(arid1a)]\n",
    "arid1a_nomutated = df.columns[~df.columns.isin(arid1a)]\n",
    "\n",
    "CDH1_mutated = df.columns[df.columns.isin(CDH1)]\n",
    "CDH1_nomutated = df.columns[~df.columns.isin(CDH1)]\n",
    "\n",
    "RHOA_mutated = df.columns[df.columns.isin(RHOA)]\n",
    "RHOA_nomutated = df.columns[~df.columns.isin(RHOA)]\n",
    "\n",
    "arid1a_mt_df = df[arid1a_mutated]\n",
    "arid1a_wt_df = df[arid1a_nomutated]\n",
    "\n",
    "CDH1_mt_df = df[CDH1_mutated]\n",
    "CDH1_wt_df = df[CDH1_nomutated]\n",
    "\n",
    "RHOA_mt_df = df[RHOA_mutated]\n",
    "RHOA_wt_df = df[RHOA_nomutated]\n",
    "\n",
    "#wildtype 대비 mutation이 높게 발현된것들 선별하기 위한 비교 데이터프레임\n",
    "CDH1_median_df = pd.DataFrame({\"wildtype\":CDH1_wt_df.median(axis=1),\"mutation\":CDH1_mt_df.median(axis=1)})\n",
    "CDH1_median_df['mt>0&wt<0'] = CDH1_median_df.apply(lambda x: \"o\" if (x['wildtype'] < 0) and (x['mutation'] > 0) else \"x\", axis=1)\n",
    "\n",
    "#wildtype 대비 mutation이 높게 발현된것들 선별하기 위한 비교 데이터프레임\n",
    "RHOA_median_df = pd.DataFrame({\"wildtype\":RHOA_wt_df.median(axis=1),\"mutation\":RHOA_mt_df.median(axis=1)})\n",
    "RHOA_median_df['mt>0&wt<0'] = RHOA_median_df.apply(lambda x: \"o\" if (x['wildtype'] < 0) and (x['mutation'] > 0) else \"x\", axis=1)\n",
    "\n",
    "#wildtype 대비 mutation이 높게 발현된것들 선별하기 위한 비교 데이터프레임\n",
    "arid1a_median_df = pd.DataFrame({\"wildtype\":arid1a_wt_df.median(axis=1),\"mutation\":arid1a_mt_df.median(axis=1)})\n",
    "arid1a_median_df['mt>0&wt<0'] = arid1a_median_df.apply(lambda x: \"o\" if (x['wildtype'] < 0) and (x['mutation'] > 0) else \"x\", axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1c1f2f1-1c5b-4c6b-b1c3-6cbd099c1370",
   "metadata": {},
   "outputs": [],
   "source": [
    "#mannwhitneyu = wilcoxon sum test 진행\n",
    "RHOA_wcxs_df = pd.DataFrame(columns=[\"col\",\"statistic\",\"pvalue\"])\n",
    "CDH1_wcxs_df = pd.DataFrame(columns=[\"col\",\"statistic\",\"pvalue\"])\n",
    "arid1a_wcxs_df = pd.DataFrame(columns=[\"col\",\"statistic\",\"pvalue\"])\n",
    "\n",
    "for idx in CDH1_wt_df.index:\n",
    "    wt_data = CDH1_wt_df.loc[idx].dropna().values\n",
    "    mt_data = CDH1_mt_df.loc[idx].dropna().values\n",
    "\n",
    "    if wt_data.size > 0 and mt_data.size > 0:\n",
    "        statistic, p_value = mannwhitneyu(wt_data, mt_data)\n",
    "        result = pd.DataFrame([[idx, statistic, p_value]], columns=[\"col\", \"statistic\", \"pvalue\"])\n",
    "        if not result.isnull().all().all() and not result.empty:\n",
    "            CDH1_wcxs_df = pd.concat([CDH1_wcxs_df, result], ignore_index=True)\n",
    "\n",
    "for idx in arid1a_wt_df.index:\n",
    "    wt_data = arid1a_wt_df.loc[idx].dropna().values\n",
    "    mt_data = arid1a_mt_df.loc[idx].dropna().values\n",
    "\n",
    "    if wt_data.size > 0 and mt_data.size > 0:\n",
    "        statistic, p_value = mannwhitneyu(wt_data, mt_data)\n",
    "        result = pd.DataFrame([[idx, statistic, p_value]], columns=[\"col\", \"statistic\", \"pvalue\"])\n",
    "        if not result.isnull().all().all() and not result.empty:\n",
    "            arid1a_wcxs_df = pd.concat([arid1a_wcxs_df, result], ignore_index=True)\n",
    "\n",
    "for idx in RHOA_wt_df.index:\n",
    "    wt_data = RHOA_wt_df.loc[idx].dropna().values\n",
    "    mt_data = RHOA_mt_df.loc[idx].dropna().values\n",
    "\n",
    "    if wt_data.size > 0 and mt_data.size > 0:\n",
    "        statistic, p_value = mannwhitneyu(wt_data, mt_data)\n",
    "        result = pd.DataFrame([[idx, statistic, p_value]], columns=[\"col\", \"statistic\", \"pvalue\"])\n",
    "        if not result.isnull().all().all() and not result.empty:\n",
    "            RHOA_wcxs_df = pd.concat([RHOA_wcxs_df, result] ,ignore_index=True)\n",
    "\n",
    "#p값 기준 0.05보다 작은 peptide 선별\n",
    "RHOA_pval_005 = RHOA_wcxs_df[RHOA_wcxs_df.pvalue < 0.05].col.to_list()\n",
    "CDH1_pval_005 = CDH1_wcxs_df[CDH1_wcxs_df.pvalue < 0.05].col.to_list()\n",
    "arid1a_pval_005 = arid1a_wcxs_df[arid1a_wcxs_df.pvalue < 0.05].col.to_list()\n",
    "\n",
    "#median값이 mutation에서 0보다 크고, wildtype에서 0보다 작은 peptide 선별\n",
    "RHOA_median_mt_over0_wt_under0 =RHOA_median_df[RHOA_median_df['mt>0&wt<0'] == \"o\"].index.to_list()\n",
    "CDH1_median_mt_over0_wt_under0 =CDH1_median_df[CDH1_median_df['mt>0&wt<0'] == \"o\"].index.to_list()\n",
    "arid1a_median_mt_over0_wt_under0 =arid1a_median_df[arid1a_median_df['mt>0&wt<0'] == \"o\"].index.to_list()\n",
    "\n",
    "#median, p값 교집합 기준 유의미한 peptide 선별\n",
    "RHOA_sig = list(set(RHOA_median_mt_over0_wt_under0)&set(RHOA_pval_005))\n",
    "CDH1_sig = list(set(CDH1_median_mt_over0_wt_under0)&set(CDH1_pval_005))\n",
    "arid1a_sig = list(set(arid1a_median_mt_over0_wt_under0)&set(arid1a_pval_005))\n",
    "\n",
    "RHOA_final_input = df[df.index.isin(RHOA_sig)]\n",
    "CDH1_final_input = df[df.index.isin(CDH1_sig)]\n",
    "arid1a_final_input = df[df.index.isin(arid1a_sig)]\n",
    "\n",
    "#heatmap에서 색깔로 mutation과 wildtype을 구분하기 위해 mutation과 wildtype끼리 모아서 concat\n",
    "RHOA_wt = RHOA_final_input[RHOA_final_input.columns[~RHOA_final_input.columns.isin(RHOA)]]\n",
    "RHOA_mt = RHOA_final_input[RHOA_final_input.columns[RHOA_final_input.columns.isin(RHOA)]]\n",
    "RHOA_final = pd.concat([RHOA_mt, RHOA_wt],axis=1)\n",
    "\n",
    "CDH1_wt = CDH1_final_input[CDH1_final_input.columns[~CDH1_final_input.columns.isin(CDH1)]]\n",
    "CDH1_mt = CDH1_final_input[CDH1_final_input.columns[CDH1_final_input.columns.isin(CDH1)]]\n",
    "CDH1_final = pd.concat([CDH1_mt, CDH1_wt],axis=1)\n",
    "\n",
    "arid1a_wt = arid1a_final_input[arid1a_final_input.columns[~arid1a_final_input.columns.isin(arid1a)]]\n",
    "arid1a_mt = arid1a_final_input[arid1a_final_input.columns[arid1a_final_input.columns.isin(arid1a)]]\n",
    "arid1a_final = pd.concat([arid1a_mt, arid1a_wt],axis=1)\n",
    "\n",
    "#heatmap을 그리기 위해 -1 ~ 1 로 스케일링\n",
    "from sklearn.preprocessing import MaxAbsScaler\n",
    "scaler = MaxAbsScaler()\n",
    "RHOA_final_cols = RHOA_final.columns\n",
    "RHOA_final_scaler = RHOA_final.copy()\n",
    "RHOA_final_scaler[RHOA_final_cols] = scaler.fit_transform(RHOA_final_scaler[RHOA_final_cols])\n",
    "\n",
    "from sklearn.preprocessing import MaxAbsScaler\n",
    "scaler = MaxAbsScaler()\n",
    "CDH1_final_cols = CDH1_final.columns\n",
    "CDH1_final_scaler = CDH1_final.copy()\n",
    "CDH1_final_scaler[CDH1_final_cols] = scaler.fit_transform(CDH1_final_scaler[CDH1_final_cols])\n",
    "\n",
    "from sklearn.preprocessing import MaxAbsScaler\n",
    "scaler = MaxAbsScaler()\n",
    "arid1a_final_cols = arid1a_final.columns\n",
    "arid1a_final_scaler = arid1a_final.copy()\n",
    "arid1a_final_scaler[arid1a_final_cols] = scaler.fit_transform(arid1a_final_scaler[arid1a_final_cols])\n",
    "\n",
    "RHOA_final_scaler.fillna(0).to_csv(\"./RHOA_heatmapinput.csv\",index=False)\n",
    "CDH1_final_scaler.fillna(0).to_csv(\"./CDH1_heatmapinput.csv\",index=False)\n",
    "arid1a_final_scaler.fillna(0).to_csv(\"./arid1a_heatmapinput.csv\",index=False)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc7c895f-a6c3-44ec-a58f-5095eeea8c3f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#patient metadata: annotation bar inpt생성\n",
    "patient_meta = pd.read_csv(\"metadata.txt\",sep=\"\\t\")\n",
    "\n",
    "#mutation 표시를 위해 해당 gene이 mutation된 환자들과 wildtype군 구분\n",
    "RHOA = [\"N111T112\",....]\n",
    "CDH1 = [\"N115T116\",....]\n",
    "arid1a = [\"N117T118\",....]\n",
    "\n",
    "#heatmap input과 동일하게 patient id 형식 맞춰주기\n",
    "patient_meta[\"Normal\"]=\"N\"+patient_meta[\"Normal\"].str[:-1]\n",
    "patient_meta[\"Tumor\"]=\"T\"+patient_meta[\"Tumor\"].str[:-1]\n",
    "patient_meta[\"sampleID\"]=patient_meta[\"Normal\"]+patient_meta[\"Tumor\"]\n",
    "\n",
    "#metadata중 annotation bar을 그릴 컬럼만 선택\n",
    "patient_meta_selected = patient_meta[[\"sampleID\",\"EBV\",\"MSI\",\"Gender\",\"Histology (Lauren)\"]]\n",
    "patient_meta_selected = patient_meta_selected.rename(columns={\"Histology (Lauren)\":\"Histology\"})\n",
    "\n",
    "patient_meta_selected['Histology'] = patient_meta_selected['Histology'].fillna(\"Others\")  #수정\n",
    "patient_meta_selected['EBV'] = patient_meta_selected['EBV'].replace(\"Negative\", \"EBV-\")\n",
    "patient_meta_selected['EBV'] = patient_meta_selected['EBV'].replace(\"EBV(PIK3CAmut)\", \"EBV+\")\n",
    "patient_meta_selected['EBV'] = patient_meta_selected['EBV'].replace(\"EBV\", \"EBV+\")\n",
    "patient_meta_selected['EBV'] = patient_meta_selected['EBV'].replace(\"EBV(PIK3CAwt)\", \"EBV+\")\n",
    "\n",
    "arid1a_patient_meta_selected = patient_meta_selected.copy()\n",
    "rhoa_patient_meta_selected = patient_meta_selected.copy()\n",
    "cdh1_patient_meta_selected = patient_meta_selected.copy()\n",
    "\n",
    "#arid1a에 mutation 여부에 대한 환자 정보 patient_meta_selected에 컬럼 추가하기\n",
    "arid1a_patient_meta_selected[\"Mutation\"] = arid1a_patient_meta_selected.sampleID.apply(lambda x: \"Mut\" if x in arid1a else \"WT\")\n",
    "\n",
    "#CDH1에 mutation 여부에 대한 환자 정보 patient_meta_selected에 컬럼 추가하기\n",
    "cdh1_patient_meta_selected[\"Mutation\"] = cdh1_patient_meta_selected.sampleID.apply(lambda x: \"Mut\" if x in CDH1 else \"WT\")\n",
    "\n",
    "#RHOA에 mutation 여부에 대한 환자 정보 patient_meta_selected에 컬럼 추가하기\n",
    "rhoa_patient_meta_selected[\"Mutation\"] = rhoa_patient_meta_selected.sampleID.apply(lambda x: \"Mut\" if x in RHOA else \"WT\")\n",
    "\n",
    "#mutation과 wildtype을 구분해서 heatmap annotation을 그리기 위해 dataframe sorting\n",
    "arid1a_patient_mt = arid1a_patient_meta_selected[arid1a_patient_meta_selected.sampleID.isin(arid1a)]\n",
    "arid1a_patient_wt = arid1a_patient_meta_selected[~arid1a_patient_meta_selected.sampleID.isin(arid1a)]\n",
    "arid1a_patient_meta_sorted = pd.concat([arid1a_patient_mt,arid1a_patient_wt],axis=0).set_index(\"sampleID\")\n",
    "arid1a_patient_meta_sorted.to_csv(\"./arid1a_patient_meta.csv\",index=False)\n",
    "\n",
    "cdh1_patient_mt = cdh1_patient_meta_selected[cdh1_patient_meta_selected.sampleID.isin(CDH1)]\n",
    "cdh1_patient_wt = cdh1_patient_meta_selected[~cdh1_patient_meta_selected.sampleID.isin(CDH1)]\n",
    "cdh1_patient_meta_sorted = pd.concat([cdh1_patient_mt,cdh1_patient_wt],axis=0).set_index(\"sampleID\")\n",
    "cdh1_patient_meta_sorted.to_csv(\"./CDH1_patient_meta.csv\",index=False)\n",
    "\n",
    "rhoa_patient_mt = rhoa_patient_meta_selected[rhoa_patient_meta_selected.sampleID.isin(RHOA)]\n",
    "rhoa_patient_wt = rhoa_patient_meta_selected[~rhoa_patient_meta_selected.sampleID.isin(RHOA)]\n",
    "rhoa_patient_meta_sorted = pd.concat([rhoa_patient_mt,rhoa_patient_wt],axis=0).set_index(\"sampleID\")\n",
    "rhoa_patient_meta_sorted.to_csv(\"./RHOA_patient_meta.csv\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9651cc3-2ef4-46fa-8a22-ae4d5a87a77a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# R STUDIO visualization preparation\n",
    "import pandas as pd\n",
    "\n",
    "# CSV 파일 읽기\n",
    "file_path = '(gene_name)_heatmapinput_P.csv' \n",
    "df = pd.read_csv(file_path)\n",
    "\n",
    "# 모든 값에 1.5 곱하기 (숫자 열만)\n",
    "df_scaled = df.select_dtypes(include=['float64', 'int64']) * 1.5\n",
    "\n",
    "# 원래 데이터프레임에서 숫자가 아닌 열을 유지하고 다시 결합\n",
    "df[df_scaled.columns] = df_scaled\n",
    "\n",
    "# 새로운 CSV 파일로 저장\n",
    "df.to_csv('(gene_name)_heatmapinput.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a467e13a-8cd6-49a3-aa6d-187c175691e1",
   "metadata": {},
   "source": [
    "# 02_Phosphopeptides_Pathway\n",
    "### --------------------\n",
    "### --------------------\n",
    "### --------------------"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30ba4e1c-9c3b-4e90-b85e-c52045a8f69c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ad2530a-f06e-4b12-97c7-1a3170930101",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#consensuspathdb 결과값을 input으로 받아오기\n",
    "arid1a_df = pd.read_csv(\"ORA_results_ARID1A.tab\",sep=\"\\t\")\n",
    "rhoa_df = pd.read_csv(\"ORA_results_RHOA.tab\",sep=\"\\t\")\n",
    "cdh1_df = pd.read_csv(\"ORA_results_CDH1.tab\",sep=\"\\t\")\n",
    "\n",
    "#필요한 컬럼만 가져오기\n",
    "arid1a_df = arid1a_df[[\"pathway\",\"q-value\"]]\n",
    "cdh1_df = cdh1_df[[\"pathway\",\"q-value\"]]\n",
    "rhoa_df = rhoa_df[[\"pathway\",\"q-value\"]]\n",
    "\n",
    "#log10취하기\n",
    "arid1a_df[\"log10_qvalue\"]=np.log10(arid1a_df[\"q-value\"])\n",
    "rhoa_df[\"log10_qvalue\"]=np.log10(rhoa_df[\"q-value\"])\n",
    "cdh1_df[\"log10_qvalue\"]=np.log10(cdh1_df[\"q-value\"])\n",
    "\n",
    "#abs로 x축 양수로 만들기\n",
    "arid1a_df[\"-log10(qvalue)\"] = abs(arid1a_df[\"log10_qvalue\"])\n",
    "\n",
    "rhoa_df[\"-log10(qvalue)\"] = abs(rhoa_df[\"log10_qvalue\"])\n",
    "\n",
    "cdh1_df[\"-log10(qvalue)\"] = abs(cdh1_df[\"log10_qvalue\"])\n",
    "\n",
    "#pathway name과 qvalue 컬럼만 가져오기\n",
    "arid1a_q = arid1a_df[[\"pathway\",\"-log10(qvalue)\"]]\n",
    "cdh1_q = cdh1_df[[\"pathway\",\"-log10(qvalue)\"]]\n",
    "rhoa_q = rhoa_df[[\"pathway\",\"-log10(qvalue)\"]]\n",
    "\n",
    "arid1a_q_sort = arid1a_q.sort_values(by=\"-log10(qvalue)\", ascending=False)\n",
    "cdh1_q_sort = cdh1_q.sort_values(by=\"-log10(qvalue)\", ascending=False)\n",
    "rhoa_q_sort = rhoa_q.sort_values(by=\"-log10(qvalue)\", ascending=False)\n",
    "\n",
    "#top10 가져오기\n",
    "cdh1_top10 = cdh1_q_sort.iloc[:10,:]\n",
    "arid1a_top10 = arid1a_q_sort.iloc[:10,:]\n",
    "rhoa_top10 = rhoa_q_sort.iloc[:10,:]\n",
    "\n",
    "#세가지 gene에 대한 pathway의 합집합 생성\n",
    "all_df = pd.DataFrame(list(arid1a_top10[\"pathway\"].to_list()+cdh1_top10[\"pathway\"].to_list()+rhoa_top10[\"pathway\"].to_list()),columns=[\"pathway\"])\n",
    "\n",
    "#세가지 gene에 대해 all_df 결과에 있는 값들 가져오기\n",
    "cdh1_select = cdh1_q[cdh1_q[\"pathway\"].isin(all_df[\"pathway\"])]\n",
    "arid1a_select = arid1a_q[arid1a_q[\"pathway\"].isin(all_df[\"pathway\"])]\n",
    "rhoa_select = rhoa_q[rhoa_q[\"pathway\"].isin(all_df[\"pathway\"])]\n",
    "\n",
    "#3개 유전자 결과 병합\n",
    "all_df_cdh1 =pd.merge(cdh1_select, all_df, on=\"pathway\", how = \"outer\").rename(columns={\"-log10(qvalue)\":\"CDH1\"})\n",
    "all_df_rhoa =pd.merge(rhoa_select, all_df_cdh1, on=\"pathway\", how = \"outer\").rename(columns={\"-log10(qvalue)\":\"RHOA\"})\n",
    "all_df_arid1a =pd.merge(arid1a_select, all_df_rhoa, on=\"pathway\", how = \"outer\").rename(columns={\"-log10(qvalue)\":\"ARID1A\"})\n",
    "\n",
    "#결과 저장\n",
    "all_df_arid1a.drop_duplicates(\"pathway\").to_csv(\"-log10pvalue_pathway.csv\",index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
